src/libre/ast_analysis.c

```diff
@@ -64,38 +86,95 @@ analysis_iter(struct ast_expr *n)
 	case AST_EXPR_LITERAL:
 	case AST_EXPR_CODEPOINT:
 	case AST_EXPR_RANGE:
-		/* no special handling */
+		set_flags(n, AST_FLAG_CAN_CONSUME | AST_FLAG_ALWAYS_CONSUMES);
 		break;
 
 	case AST_EXPR_CONCAT: {
 		size_t i;
+		int any_can_consume = 0;
+		int any_always_consumes = 0;
+		int all_nullable = 1;
 
 		for (i = 0; i < n->u.concat.count; i++) {
+			struct ast_expr *child = n->u.concat.n[i];
 			if (is_nullable(n)) {
-				set_flags(n->u.concat.n[i], AST_FLAG_NULLABLE);
+				LOG(3 - LOG_FIRST_ANALYSIS, "%p: setting child %p nullable\n",
+				    (void *)n, (void *)child);
+				set_flags(child, AST_FLAG_NULLABLE);
+			}
+			analysis_iter(child);
+			if (can_consume_input(child)) {
+				any_can_consume = 1;
 			}
-			analysis_iter(n->u.concat.n[i]);
+			if (always_consumes_input(child)) {
+				any_always_consumes = 1;
+			}
+
+			if (!is_nullable(child)) {
+				all_nullable = 0;
+			}
+		}
+
+		if (any_can_consume) {
+			LOG(3 - LOG_FIRST_ANALYSIS, "%p: any_can_consume -> setting CAN_CONSUME\n",
+			    (void *)n);
+			set_flags(n, AST_FLAG_CAN_CONSUME);
+		}
+
+		if (any_always_consumes) {
+			LOG(3 - LOG_FIRST_ANALYSIS, "%p: any_always_consumes -> setting ALWAYS_CONSUMES\n",
+			    (void *)n);
+			assert(any_can_consume);
+			set_flags(n, AST_FLAG_ALWAYS_CONSUMES);
+		}
+		if (n->u.concat.count > 0 && all_nullable) {
+			set_flags(n, AST_FLAG_NULLABLE);
 		}
 
 		break;
 	}
 
 	case AST_EXPR_ALT: {
 		size_t i;
+		int any_can_consume = 0;
+		size_t always_consumes_count = 0;
 
 		for (i = 0; i < n->u.alt.count; i++) {
-			analysis_iter(n->u.alt.n[i]);
+			struct ast_expr *child = n->u.alt.n[i];
+			analysis_iter(child);
+
 			/* spread nullability upward */
-			if (is_nullable(n->u.alt.n[i])) {
+			if (is_nullable(child)) {
+				LOG(3 - LOG_FIRST_ANALYSIS, "%p: child %p is nullable, spreading upward\n",
+				    (void *)child, (void *)n);
 				set_flags(n, AST_FLAG_NULLABLE);
 			}
-		}
 
-		if (is_nullable(n)) { /* spread nullability downward */
-			for (i = 0; i < n->u.alt.count; i++) {
-				set_flags(n->u.alt.n[i], AST_FLAG_NULLABLE);
+			if (can_consume_input(child)) {
+				any_can_consume = 1;
+			}
+			if (always_consumes_input(child)) {
+				always_consumes_count++;
 			}
 		}
+
+		if (any_can_consume) {
+			LOG(3 - LOG_FIRST_ANALYSIS,
+			    "%p: any_can_consume -> setting CAN_CONSUME\n", (void *)n);
+			set_flags(n, AST_FLAG_CAN_CONSUME);
+		}
+
+		/* FIXME: This should not be invalidated by unsatisfiable nodes without the
+		 * AST_FLAG_ALWAYS_CONSUMES flag set, but since unsatisfiability derives
+		 * from needing to consume input in a way that contradicts anchoring, it
+		 * may not be possible to produce a case where an alt subtree is
+		 * unsatisfiable but does not set AST_FLAG_ALWAYS_CONSUMES. */
```

Comment:
Would it make sense to reject compilation for this situation, just in case it does occur? I'm taking the "it may not be possible" part to mean that we don't have test coverage for this situation.

Reply:
I don't think it's possible here because it relies on information that won't be available until an analysis pass that has to run after this one completes, but I also haven't been able to find an input that produces incorrect results due to that situation.

I think rejecting it would be for the best, but a later integration PR will rework error handling for unsupported input -- there currently isn't a way to explicitly reject a regex because of unsupported interactions between captures and other features -- so it should be easier to address later on.